src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -249,6 +249,172 @@ public static Path of(URI uri) {
      */
     Path getFileName();
 
+    /**
+     * Returns the file extension of this path's file name as a {@code String}.
+     * The extension is acquired from this {@code Path} by obtaining the
+     * {@linkplain #getFileName file name element}, deriving its {@linkplain
+     * #toString string representation}, and then extracting a substring
+     * determined by the position of the last period character ('.', U+002E
+     * FULL STOP) within the file name string. If the file name element is
+     * {@code null}, or if the file name string does not contain a period
+     * character, or if the last character in the file name string is a period,
+     * or if the prefix of the file name string is a sequence of periods
+     * including the last period, then the extension is
+     * {@linkplain String#isEmpty empty}.
+     *
+     * @implSpec
+     * The default implementation is in most cases equivalent for this path to:
+     * {@snippet lang="java" :
+     *     int lastPeriod = fileNameString.lastIndexOf('.');
+     *     return lastPeriod <= 0
+     *        ? ""
+     *        : fileNameString.substring(lastPeriod + 1);
+     * }
+     *
+     * @return  the file name extension of this path, which might be the
+     *          empty string
+     *
+     * @see #withExtension
+     * @see #withoutExtension
+     *
+     * @since 23
+     */
+    default String getExtension() {
+        Path fileName = getFileName();
+        if (fileName == null)
+            return "";
+
+        String fileNameString = fileName.toString();
+        int lastPeriodIndex = fileNameString.lastIndexOf('.');
+
+        if (lastPeriodIndex <= 0)
+            return "";
+
+        boolean prefixIsPeriods = true;
+        for (int idx = lastPeriodIndex - 1; idx >= 0; idx--) {
+            if (fileNameString.charAt(idx) != '.') {
+                prefixIsPeriods = false;
+                break;
+            }
+        }
+        if (prefixIsPeriods)
+            return "";
+
+        // If lastPeriodIndex == fileNameString.length() - 1 then "" is returned
+        return fileNameString.substring(lastPeriodIndex + 1);
+    }
+
+    /**
+     * Returns a copy of this {@code Path} with the file name extension removed.
+     * If this path has no extension, then the path is returned unchanged.
+     *
+     * <p> A compound extension may be replaced by invoking this method and
+     * then the {@linkplain #withExtension withExtension} method on the result.
+     * For example, to convert the {@code "tar.gz"} compound extension
+     * to {@code "zip"}:
+     * {@snippet lang="java" :
+     *    Path p = Path.of("archive.tar.gz");
+     *    Path q = p.withoutExtension().withExtension("zip");
+     * }
+     * where {@code q.toString()} would return {@code "archive.zip"}.
+     *
+     * @apiNote
+     * This method must satisfy the invariant:
+     * {@snippet lang="java" :
+     *     assert equals(Path.of(withoutExtension().toString()
```

Comment:
I'm an author of an HTTP filesystem provider and would like to use these new methods since they address longstanding pain points working with Paths.  However, my usage wouldn't be able to satisfy this requirement. 

Here's an example of on our path strings:  `http://example.com/subfolder/file.txt?language=english`

In this case, the filename is `file.txt` and the query parameters (`?language=english`) are not considered part of the filename.
This invariant doesn't work for such a path.
```
withoutExtension() == "http://example.com/subfolder/file?language=english"
getExtension() == "txt"
```
So it would end up looking like:
```
Path.of("http://example.com/subfolder/file?language=english" + ".txt"). ==  "http://example.com/subfolder/file?language=englis.txt"
```

I think the invariant might be better written something like this:

```
path.equals(path.withExtension(path.getExtension())
```

As an aside, it's not generally true that `Path.of(path.toString()).equals(path)` so if you are going to require that you have to specify it only applies to the default FileSystemProvider.
```

To construct more general Path implementations you have to use the `Path.of(URI)` method instead of `Path.of(String)` so it would probably be better to refer to that.

Reply:
As for the example "http://example.com/subfolder/file.txt?language=english", this is what I see at present:
```
jshell> Path p = Path.of("http://example.com/subfolder/file.txt?language=english")
p ==> http:/example.com/subfolder/file.txt?language=english

jshell> p.withoutExtension()
$2 ==> http:/example.com/subfolder/file

jshell> p.getExtension()
$3 ==> "txt?language=english"

jshell> p.withoutExtension().toString() + "." + p.getExtension()
$4 ==> "http:/example.com/subfolder/file.txt?language=english"

jshell> p.withExtension(p.getExtension())
$5 ==> http:/example.com/subfolder/file.txt?language=english
```
The two forms of the invariant appear satisfied.

Good point about non-default providers.

Reply:
I would think that manipulations of filename components of a URI/URL would require extracting/converting to a file path first. Path applies to file system names. One could imagine a FileSystem whose Paths were URIs and then yes it could/should implement the file extension related methods to manipulate the file extension and produce valid URIs.

Reply:
Yes, our software uses a number of alternative file systems in order to treat file access uniformly across different domains.   One for google cloud storage, one for amazon, one for arbitrary http paths, etc.   They're constructed using the `Path.of(URI)` mechanism which matches filesystems by URI scheme.  I just wanted to point out that invariant around handling of toString() wouldn't hold for non-default implementations of Path so maybe it should be stated more generally in terms of path operations instead of in terms of string concatenation.

Reply:
It make sense to express the behavior in terms of Path though the default implementation needs to be able to take advantage of the mapping between strings and Paths to be able to describe and define the behavior that is consistent across filesystems.